src/server/glib_main_loop.cpp

```diff
@@ -398,51 +398,72 @@ void mir::GLibMainLoop::reprocess_all_sources()
 
 void mir::GLibMainLoop::run_with_context_as_thread_default(std::function<void()> const& code)
 {
-    auto promise = std::make_shared<std::promise<void>>();
-    auto done = promise->get_future();
+    if (g_main_context_acquire(main_context))
+    {
+        /* We have the main context, so just do the thing. */
+        g_main_context_push_thread_default(main_context);
 
-    enqueue_with_guaranteed_execution(
-        [this, promise, &code]()
+        try
+        {
+            code();
+        }
+        catch (...)
         {
-            /*
-             * There are three possible states here:
-             * 1)   We've been dispatched from the main loop; the thread already owns main_context,
-             *      so we can acquire it.
-             * 2)   The main loop is not running, so we've been immediately dispatched from the calling
-             *      thread. The main loop is not running, so we can successfully acquire main_context
-             *      immediately.
-             * 3)   The main loop *is* running, but is currently processing a stop() request.
-             *      In this case, we've been immediately dispatched from the calling thread, but
-             *      cannot acquire main_context until stop() has completed.
-             *
-             *      There *is* a GLib API that would be helpful here; g_main_context_wait().
-             *      however, that is (silently) deprecated, and spews a GLib-CRITICAL warning
-             *      if you use it.
-             *
-             *      This case is unlikely, and is should be rapidly transient, so just busy-wait :(
-             */
-            {
-                std::lock_guard<std::mutex> lock{run_on_halt_mutex};
-                while (!g_main_context_acquire(main_context))
-                    std::this_thread::yield();
-            }
-
-            // We now own the main_context, so this will work.
-            g_main_context_push_thread_default(main_context);
-            try
-            {
-                code();
-                promise->set_value();
-            }
-            catch(...)
-            {
-                promise->set_exception(std::current_exception());
-            }
             g_main_context_pop_thread_default(main_context);
             g_main_context_release(main_context);
-        });
+            throw;
+        }
+        g_main_context_pop_thread_default(main_context);
+        g_main_context_release(main_context);
```

Comment:
This is very nearly `mir::raii`able, but not quite, and it didn't seem worth the effort of doing something custom.